pkg/kubelet/kubelet.go

```diff
@@ -3282,7 +3311,21 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
 
 func (kl *Kubelet) generateAPIPodStatus(pod *api.Pod, podStatus *kubecontainer.PodStatus) api.PodStatus {
 	glog.V(3).Infof("Generating status for %q", format.Pod(pod))
+
+	// any pod that requires eviction must be added to the list to sync.
+	attrs := &lifecycle.PodEvictAttributes{Pod: pod}
+	for _, podEvictor := range kl.podEvictors {
+		if result := podEvictor.Evict(attrs); result.Evict {
```

Comment:
@vishh - the decision to evict has been made but the actual eviction has
not yet completed.  The getPodsToSync call happens every second, and it
polls each pod evictor to see if each pod needs to be evicted.  If so, it
adds it to the pod to sync list, the syncPod function is called in turn to
then generate the pod status and the pod status is updated to failed.  The
pod by virtue of being failed is then cleaned up in HandlePodCleanups. This
is the same basic flow for how active deadline works as well just
formalized.  A pod worker could also have discovered that the pod is failed
when it calls generate pod status but the current code path does everything
through get pods to sync flow.

Can we have a brief hangout tomorrow to discuss?  I'm open to other
implementation strategies, but this has worked well for me.

On Monday, April 18, 2016, Vish Kannan <notifications@github.com> wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/24344#discussion_r60146006>
> :
>
> > @@ -3282,7 +3311,21 @@ func GetPhase(spec *api.PodSpec, info []api.ContainerStatus) api.PodPhase {
> >
> >  func (kl *Kubelet) generateAPIPodStatus(pod *api.Pod, podStatus *kubecontainer.PodStatus) api.PodStatus {
> >  	glog.V(3).Infof("Generating status for %q", format.Pod(pod))
> > +
> > +	// any pod that requires eviction must be added to the list to sync.
> > +	attrs := &lifecycle.PodEvictAttributes{Pod: pod}
> > +	for _, podEvictor := range kl.podEvictors {
> > +		if result := podEvictor.Evict(attrs); result.Evict {
>
> @derekwaynecarr <https://github.com/derekwaynecarr> By the time we get
> here, the eviction has already taken place right?
> Is that the motivation for the logic above
> <https://github.com/kubernetes/kubernetes/pull/24344/files#diff-bf28da68f62a8df6e99e447c4351122dR2056>
> as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24344/files/d2c9b6b8abf00d3e41d29e53c5de9686a1f5f0d1#r60146006>
>